### PR TITLE
Lido wstETH Plugin for Base

### PR DESCRIPTION
**Pull request by akshatmittal**

### CODE REVIEW

**Review by tbrent — APPROVED** [review]
LGTM.

**Comment by tbrent** [comment]
I'm looking forward to the backtest so we can confirm this chainlink feed works the way we think it does.

File: test/plugins/individual-collateral/lido/L2LidoStakedEthTestSuite.test.ts [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+  maxTradeVolume: MAX_TRADE_VOL,
+  defaultThreshold: DEFAULT_THRESHOLD,
+  delayUntilDefault: DELAY_UNTIL_DEFAULT,
+  targetPerRefChainlinkFeed: BASE_PRICE_FEEDS.stETH_ETH,
+  targetPerRefChainlinkTimeout: BASE_FEEDS_TIMEOUT.stETH_ETH,
+  uoaPerTargetChainlinkFeed: BASE_PRICE_FEEDS.ETH_USD,
+  uoaPerTargetChainlinkTimeout: BASE_FEEDS_TIMEOUT.ETH_USD,
+  refPerTokenChainlinkFeed: BASE_PRICE_FEEDS.wstETH_stETH,
+  refPerTokenChainlinkTimeout: BASE_FEEDS_TIMEOUT.wstETH_stETH,
+  revenueHiding: fp('0'), // TODO: how much?
```
[the remark this answers is not in the document]
Comment (tbrent) [reply test/plugins/individual-collateral/lido/L2LidoStakedEthTestSuite.test.ts:72]: Actually, this does need revenue hiding. 0.01%, or fp('1e-4'), to be consistent with mainnet

File: test/plugins/individual-collateral/lido/constants.ts [hunk 2]
```diff
@@ -17,5 +16,21 @@ export const ORACLE_ERROR = fp('0.005')
...
+  stETH_ETH: '0xf586d0728a47229e747d824a939000Cf21dEF5A0',
+  ETH_USD: '0x71041dddad3595F9CEd3DcCFBe3D1F4b0a16Bb70',
+  wstETH_stETH: '0xB88BAc61a4Ca37C43a3725912B1f472c9A5bc061',
+}
+export const BASE_FEEDS_TIMEOUT = {
+  stETH_ETH: bn(86400),
+  ETH_USD: bn(1200), // yep, that's correct
+  wstETH_stETH: bn(86400),
+}
+export const BASE_ORACLE_ERROR = fp('0.0115') // 0.5% + 0.5% + 0.15%
```
[the remark this answers is not in the document]
Comment (pmckelvy1) [reply test/plugins/individual-collateral/lido/constants.ts:36]: well technically slashing could occur, right? so we need a non-zero error

File: contracts/plugins/assets/lido/L2LidoStakedEthCollateral.sol [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        uint192 err = p.mul(oracleError, CEIL);
+
+        high = p + err;
+        low = p - err;
+        // assert(low <= high); obviously true just by inspection
+    }
+
+    /// @return {ref/tok} Quantity of whole reference units per whole collateral tokens
+    function underlyingRefPerTok() public view override returns (uint192) {
+        return _safeWrap(refPerTokenChainlinkFeed.price(refPerTokenChainlinkTimeout));
```
Comment (pmckelvy1) [inline contracts/plugins/assets/lido/L2LidoStakedEthCollateral.sol:95]: no need to _safeWrap

File: test/plugins/individual-collateral/lido/L2LidoStakedEthTestSuite.test.ts [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  // fortunately this syntax fails silently because our tools are terrible
+  await expect(collateral.refresh())
+
+  return collateral
+}
+
+const defaultAnswers = {
+  targetPerRefChainlinkFeed: bn('1e18'),
+  uoaPerTargetChainlinkFeed: bn('2000e8'),
+  refPerTokenChainlinkFeed: bn('1e18'), // ??
```
Comment (pmckelvy1) [inline test/plugins/individual-collateral/lido/L2LidoStakedEthTestSuite.test.ts:123]: maybe 1.15e18 just to make it more realistic?

File: test/plugins/individual-collateral/lido/L2LidoStakedEthTestSuite.test.ts [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+  Define helper functions
+*/
+
+const mintCollateralTo: MintCollateralFunc<WSTETHCollateralFixtureContext> = async (
+  ctx: WSTETHCollateralFixtureContext,
+  amount: BigNumberish,
+  user: SignerWithAddress,
+  recipient: string
+) => {
+  // TODO: Change this!
```
Comment (pmckelvy1) [inline test/plugins/individual-collateral/lido/L2LidoStakedEthTestSuite.test.ts:186]: still need to change this?

**Review by pmckelvy1 — CHANGES_REQUESTED** [review]

File: test/plugins/individual-collateral/lido/constants.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (tbrent) [reply test/plugins/individual-collateral/lido/constants.ts:36]: Correct, just commenting on the weirdness

**Comment by pmckelvy1** [comment]
lgtm

**Pull request merged**